fusedrug/data/tokenizer/injectortokenizer/injector_tokenizer.py

```diff
@@ -0,0 +1,207 @@
+from tokenizers import Encoding
+import torch
+import re
+from fuse.utils import NDict
+
+
+class InjectorTokenizer(ModularTokenizer):
+    """
+    InjectorTokenizer builds on top of ModularTokenizer.
+
+    Its purpose is to extend beyond "standard" input tokens as integers as input for a model.
+    Instead, it provides control on *vectors* that are to be used as input for a model.
+
+    Example use cases:
+    1. Providing scalars (floating point) as inputs
+    2. Providing vectors of embeddings - for example of a protein embedding
+
+    Each input "token" becomes a tensor of a defined size, and is built of:
+    1. Header
+        made of 4 floats
+        [
+            0.0 or 1.0 #is this a sentinel/mask or not
+            0.0 or 1.0 #is this a standard vocabulary token
+            0.0 or 1.0 #is this a scalar
+            0.0 or 1.0 #is this a full injected vector (e.g. an embedding)
+        ]
+    2. Content
+        the rest of each input vector is made of input_dim-4 float elements.
+
+
+    Note - in the "standard vocabulary token" - we support providing an external embeding layer (like in vanilla T5),
+        as it's part of the trained weights.
+
+    """
+
+    @staticmethod
+    def build_placeholder_meta_tokenization(
+        *,
+        sequence: str,
+        sample_dict: Optional[NDict] = None,
+    ) -> Tuple[str, List[str]]:
+        """
+        In order to avoid modifying and rewriting the logic in modular tokenizer, especially regarding padding, limitation of max length of certain sub-parts,
+         we put placeholders to make sure that the total size is known/fixed and respects the meta instructions to the modular tokenizer
+
+         Returns: a tuple with 2 elements
+         (
+            a single string with the full query containing placeholder tokens for FLOAT and VECTOR meta tokenizer parts,
+            a list of [meta-tokenizer name, data, meta-tokenizer name, data, meta-tokenizer name, data,  ...]
+         )
+        """
+        hints_and_subseq = re.split("<@TOKENIZER-TYPE=([^>]*)>", sequence)[
+            1:
+        ]  # the first element is blank - removing it
+        assert (
+            len(hints_and_subseq) > 0 and len(hints_and_subseq) % 2 == 0
+        ), f"Error: expecting leading modular tokenizer hints followed by a sequence to tokenize, got {sequence}"
+
+        with_placeholders = []
+
+        for tokenizer_type, subseq in zip(
+            hints_and_subseq[::2], hints_and_subseq[1::2]
+        ):
+            if tokenizer_type.startswith("SCALARS_"):
+                with_placeholders.append(
+                    "<@TOKENIZER-TYPE=AA>"
+                )  # won't use AA tokens, just an arbitrary one to be able to use a token like <SCALAR>
+
+                if (
+                    tokenizer_type == "SCALARS_LITERALS"
+                ):  # note: masking is only supported in literals (not in "from dict")
+                    values = subseq.split(",")
```

Comment:
yes, the scalars tokenizer require that you split them with ','
if you have an alternative you prefer do suggest.
I will add some description of the expected format in the injector files docstrings

Reply:
added docstrings with format description for both injector_tokenizer.py and injector_tokenizer_op

also renamed InjectorTokenizer to InjectorTokenizerHelpers and stopped inheriting from ModularTokenizer in it because it's misleading, as it's just 2 static helper methods.

Reply:
this is part of the docstrings I've added

```
applies a injector tokenizer

    injector tokenizer builds on top of modular tokenizer.
    its purpose is to build inputs_emb for the model (instead of input_ids)
        this allows to support more advanced inputs beyond token ids, like:
        * scalars inputs
        * embeddings vector within a single input

    supported syntax/format:

    for text following <@TOKENIZER-TYPE=SCALARS_LITERALS> supports the following format:
    ',' separated float values and/or <MASK> tokens - 
        for example: "2.7,3.99,-12.9" or "<MASK><MASK>" or "2.19,<MASK>,3.19,<MASK>"
    
    for text following <@TOKENIZER-TYPE=SCALARS_FROM_DICT> is expected to be a key to the sample NDict
        for example: "blah.boo.banana"  or "data.input.encoder_input"
        note: in SCALARS_FROM_DICT you can't describe masked scalars (outputs) you can only describe inputs

    example usage:

    encoder_input:
    <@TOKENIZER-TYPE=AA><MOLECULAR_WEIGHT_IN_SOME_UNIT><@TOKENIZER-TYPE=SCALARS_LITERALS>0.3<@TOKENIZER-TYPE=AA><BINDING_AFFINITY_NANOMOLAR><@TOKENIZER-TYPE=SCALARS_LITERALS><MASK><@TOKENIZER-TYPE=AA><SEQUENCE_NATURAL_START>ISGGDAIYSSTGRCSLGFNVRSGSTYYFLTAGICTDGATTWWANSARTTVLGTTSGSSFPNNDYGIVRYTNTTIPKDGTVGGQDITSAANATVGMAVTRRGSTTGTISGSVTALNATVNYGGGDVVYGMIRTNVCAEPGDSGGPLYSGTRAIGLTSGGSGNCSSGGTTFFQPVTEALVAYGVSVY<SEQUENCE_NATURAL_END>
    labels:
    <@TOKENIZER-TYPE=AA><MOLECULAR_WEIGHT_IN_SOME_UNIT><@TOKENIZER-TYPE=SCALARS_LITERALS>0.3<@TOKENIZER-TYPE=AA><BINDING_AFFINITY_NANOMOLAR><@TOKENIZER-TYPE=SCALARS_LITERALS>12.4<@TOKENIZER-TYPE=AA><SEQUENCE_NATURAL_START>ISGGDAIYSSTGRCSLGFNVRSGSTYYFLTAGICTDGATTWWANSARTTVLGTTSGSSFPNNDYGIVRYTNTTIPKDGTVGGQDITSAANATVGMAVTRRGSTTGTISGSVTALNATVNYGGGDVVYGMIRTNVCAEPGDSGGPLYSGTRAIGLTSGGSGNCSSGGTTFFQPVTEALVAYGVSVY<SEQUENCE_NATURAL_END>
```